[bitserializer] Update to new version 0.10

PavelKisliak opened the pull request:
Update BitSerializer library to new version 0.10

- What does your PR fix? Fixes #
This PR updates BitSerializer to new version and removes next ports: bitserializer-cpprestjson, bitserializer-pugixml, bitserializer-rapidjson, bitserializer-rapidyaml.
Now all formats will be available as features.

- Which triplets are supported/not supported? Have you updated the CI baseline?
Should be supported all triplets, except **arm** and **osx** but only when install with feature "rapidyaml-archive" (as base library ryml also does not support these triplets too).
Removed mentions of "bitserializer-rapidyaml" port from CI baseline file.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
√

PavelKisliak committed:
[BitSerializer] Update to version 0.9

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg

Commit:
[bitserializer] Update to version 0.10

PhoebeHui commented:
From CI failure log: it looks  BitSerializer requires at least gcc-8.3, it‘s using GNU 7.5.0 in our linux CI test machine, @dan-shaw, could you help have a look?

CMake Warning at /mnt/_work/1/s/scripts/buildsystems/vcpkg.cmake:157 (message):
  There are no libraries installed for the Vcpkg triplet x64-linux.
Call Stack (most recent call first):
  /mnt/_work/1/s/downloads/tools/cmake-3.17.2-linux/cmake-3.17.2-Linux-x86_64/share/cmake-3.17/Modules/CMakeDetermineSystem.cmake:93 (include)
  CMakeLists.txt:13 (project)

CMake Error at CMakeLists.txt:52 (message):
  BitSerializer requires at least gcc-8.3

PhoebeHui reviewed (approved):
Thanks for the PR!

PavelKisliak commented:
@PhoebeHui, many thanks for fast review!
I have one question. I did installing by default just "core" feature, it is in connection that consumers would not want to install all supported formats with all their dependencies. But as I understand the current VCPKG CI uses just default features set in the port. Is it possible to configure run CI with full set of features? Like "bitserializer[cpprestjson-archive,rapidjson-archive,pugixml-archive,rapidyaml-archive]"? Or be will be more right to enable all set of features by default?

cenit commented:
As of now, to test in ci a feature you have to add it to default features

PhoebeHui commented:
@PavelKisliak, if you want to test all features via CI, you can add them to default feature temporarily, after everything test pass, you can change that back.

Commit:
[bitserializer] Enable all features by default

PhoebeHui commented:
https://bitbucket.org/Pavel_Kisliak/bitserializer/src/c777bd4f5167598b9a65bef80999812634390ec6/CMakeLists.txt#lines-48
```
if (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS "8.0")
    message(FATAL_ERROR "BitSerializer requires at least clang-8.0")
endif()
if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS "8.3")
    message(FATAL_ERROR "BitSerializer requires at least gcc-8.3")
endif()
```

PavelKisliak commented:
Thank you guys, I've enabled all features, and see that CI fails on **MacOS**, I didn't see the reason (as I can't see logs with details), but I think that it is related to feature "rapidyaml-archive" which requires "ryml" library which does not support **MacOS**. As I understand there are no possibilities to disable build on CI for specific set of features.

About my check for minimal supported compilers in **Cmake**, I understand that it is not important for **install** flow, not sure, may be it is not good place for such check?

PhoebeHui commented:
@PavelKisliak, I upgraded gcc to 8.4.0 in my VM, and it still failed with following failure:

```
/home/phoebe/11683/vcpkg$ cat buildtrees/bitserializer/config-x64-linux-dbg-err.log
CMake Error at CMakeLists.txt:52 (message):
  BitSerializer requires at least gcc-8.3


/home/phoebe/11683/vcpkg$
/home/phoebe/11683/vcpkg$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/8/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 8.4.0-1ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-8/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-8 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 8.4.0 (Ubuntu 8.4.0-1ubuntu1~18.04)
```

PavelKisliak commented:
Interesting, I have CI with compilation on GCC 8:
https://dev.azure.com/real0793/BitSerializer/_build/results?buildId=460&view=logs&j=5b46442d-46e5-54d7-db8c-9561d489a46d

PavelKisliak commented:
I will remove this check and update PR a bit later.

cenit commented:
@PhoebeHui what is the output of `echo $CC` on your pc? Alternatively, what is the compiler adopted by cmake (In cmakecache)?

Commit:
[bitserializer] Update hash of tagged commit and remove "rapidyaml-archive" from default list of features

PavelKisliak commented:
I've made warning message instead of fail when compilers does not meet requirements and removed "rapidyaml-archive" from default list of features, now build is green. But I think that will be good to upgrade compiler on CI because only **GCC-8** has full support of **C++ 17**.

I see one more issue, but seems it does not related to port file (it works as expected), when install **BitSerializer** with all features:
```
vcpkg install bitserializer[cpprestjson-archive,rapidjson-archive,pugixml-archive,rapidyaml-archive]:x64-windows
```
In the output log missed one target "BitSerializer::cpprestjson-archive":
```
The package bitserializer:x64-windows provides CMake targets:

    find_package(bitserializer CONFIG REQUIRED)
    # Note: 1 target(s) were omitted.
    target_link_libraries(main PRIVATE BitSerializer::core BitSerializer::pugixml-archive BitSerializer::rapidjson-archive BitSerializer::rapidyaml-archive)
```

If install for example: `bitserializer[cpprestjson-archive,pugixml-archive]` than message about targets looks good:
```
The package bitserializer:x64-windows provides CMake targets:

    find_package(bitserializer CONFIG REQUIRED)
    target_link_libraries(main PRIVATE BitSerializer::core BitSerializer::pugixml-archive BitSerializer::rapidjson-archive BitSerializer::cpprestjson-archive)
```
Seems there are some limit for display targets.

PhoebeHui commented:
The gcc enviroment in my VM is mess, I reinstalled gcc 7.5.0 and gcc 8.4.0, 


Starting package 58/58: bitserializer:x64-linux
Building package bitserializer[core,cpprestjson-archive,pugixml-archive,rapidjson-archive]:x64-linux...
-- Downloading https://bitbucket.com/Pavel_Kisliak/BitSerializer/get/0.10.tar.gz...
-- Extracting source /home/phoebe/11683/vcpkg/downloads/Pavel_Kisliak-BitSerializer-0.10.tar.gz
-- Using source at /home/phoebe/11683/vcpkg/buildtrees/bitserializer/src/0.10-71fe0ca140
-- Configuring x64-linux-dbg
-- Configuring x64-linux-rel
-- Building x64-linux-dbg
-- Building x64-linux-rel
-- Installing: /home/phoebe/11683/vcpkg/packages/bitserializer_x64-linux/share/bitserializer/copyright
-- Performing post-build validation
-- Performing post-build validation done
Building package bitserializer[core,cpprestjson-archive,pugixml-archive,rapidjson-archive]:x64-linux... done
Installing package bitserializer[core,cpprestjson-archive,pugixml-archive,rapidjson-archive]:x64-linux...
Installing package bitserializer[core,cpprestjson-archive,pugixml-archive,rapidjson-archive]:x64-linux... done
Elapsed time for package bitserializer:x64-linux: 3.209 s

Total elapsed time: 12.11 min

The package bitserializer:x64-linux provides CMake targets:

    find_package(bitserializer CONFIG REQUIRED)
    target_link_libraries(main PRIVATE BitSerializer::core BitSerializer::pugixml-archive BitSerializer::rapidjson-archive BitSerializer::cpprestjson-archive)

PhoebeHui commented:
@dan-shaw, I noticed when build B2(boost-build), it requires C++11 feature, it failed when building with gcc 8.4.0, then I swich to gcc 7.50, everything builts successfully.

PhoebeHui commented:
@PavelKisliak, it looks 3 of 4 features(except rapidyaml-archive) test pass with all triplets in CI system, The feature 'rapidyaml-archiv failed due to ryml that doesn't support macos. 

Everything is done, right? could you please remove the default feature since the features have been test?

PhoebeHui commented:
@dan-shaw, any updates about removing ports? remove it directly or set it empty package?

PavelKisliak commented:
@PhoebeHui I think that we can merge as it is now - with enabled features. This should help to keep consistency between libraries via VCPKG CI. Would be great if VCPKG will support configure CI more precisely - allow to define set of features and which platforms they supports.

About issue with displaying targets, I see that you installed with default features set - which currently does not include "rapidyaml-archive", please try:
```
vcpkg install bitserializer[cpprestjson-archive,rapidjson-archive,pugixml-archive,rapidyaml-archive]:x64-windows
```
I see that VCPKG displays maximum 4 targets now, but BitSerializer has 5 targets (core + 4 optional features). It's not important for now, as library works fine, but looks as defect.

cenit commented:
@PavelKisliak the problem now is the upgrade path for people which already had your previous ports installed. They find themselves with vcpkg broken if you delete ports. It would be better if you just replaced the missing ports with empty ones, which simply depends on the unified one with the feature enabled

PhoebeHui commented:
You can reference ports https://github.com/microsoft/vcpkg/tree/master/ports/catch for setting the ports to empty packages.

PhoebeHui commented:
@PavelKisliak, sorry for the confuse.

vcpkg upgrade has a bug when removing a port or feature, so we set the package as empty temproraliy. Related issue https://github.com/microsoft/vcpkg/issues/11482

I think these deprecated ports should be removed when the bug fixed. but now we would accept the workaround for unblocking users.

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into bitserializer

Commit:
[bitserializer] Add alias packages for deprecated BitSerializer's libraries

PavelKisliak commented:
@PhoebeHui I've committed aliases for packages. I see that failed build of BitSerializer for arm_uwp, but I don't make changes in the main port file.

PhoebeHui commented:
on arm-uwp, it failed due to 
CMake Error at C:/agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:331 (_find_package):
  Could not find a package configuration file provided by "ryml" with any of
  the following names:

    rymlConfig.cmake
    ryml-config.cmake

  Add the installation prefix of "ryml" to CMAKE_PREFIX_PATH or set
  "ryml_DIR" to a directory containing one of the above files.  If "ryml"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  CMakeLists.txt:132 (find_package)

PhoebeHui commented:
/azp run

PhoebeHui committed:
Skip bitserializer-rapidyaml for arm and osx

PavelKisliak commented:
**Ryml** does not support **arm-uwp** and that reason why I not included **rapidyaml-archive** to default set of features. It should be not built on **arm-uwp** and bitserializer[rapidyaml-archive] too.

As I understand this error has appeared after I added empty packages (aliases). As you can see I've added to CONTROL-file (for **bitserializer-rapidyaml**) key "Supports: !(arm|arm64|osx)" and added exceptions to "ci.baseline.txt" for **bitserializer-rapidyaml** (they were the same in previous version).

Not sure, but may be port aliases forces to build main port **bitserializer** with all enabled features?
I did dependencies only for required features, for example, alias port **bitserializer-rapidjson** requires **bitserializer[rapidjson-archive]**.

PhoebeHui commented:
bitserializer[rapidyaml-archive] doesn't support arm-uwp, bitserializer-rapidyaml rely on bitserializer[rapidyaml-archive], even it marked as fail in CI baseline file, the dependency port bitserializer[rapidyaml-archive] still be load to build, so it failed here.

I try to skip them in CI testing, it seems that doesn't work.

PhoebeHui commented:
@BillyONeal, could you help confirm this issue?

I noticed that arm64-windows doesn't fail in CI testing, I expect the dependency port bitserializer[rapidyaml-archive]:arm-uwp should be excluded here since we skip bitserializer-rapidyaml:arm-uwp.

PhoebeHui commented:
Same issue on OSX, from the log, feature bitserializer[rapidyaml-archive] load to build even we set  
bitserializer-rapidyaml:x64-osx=skip in CI baseline file.

```
Starting package 70/73: bitserializer:x64-osx
Building package bitserializer[core,cpprestjson-archive,pugixml-archive,rapidjson-archive,rapidyaml-archive]:x64-osx...
Could not locate cached archive: /Users/vagrant/Data/archives/b1/b1c385d19c96f3be04a4e64b0669043bb29a2b52.zip
-- Downloading https://bitbucket.com/Pavel_Kisliak/BitSerializer/get/0.10.tar.gz...
-- Extracting source /Users/vagrant/Data/downloads/Pavel_Kisliak-BitSerializer-0.10.tar.gz
-- Using source at /Users/vagrant/Data/buildtrees/bitserializer/src/0.10-71fe0ca140
-- Configuring x64-osx-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: /Volumes/data/downloads/tools/cmake-3.17.2-osx/cmake-3.17.2-Darwin-x86_64/CMake.app/Contents/bin/cmake /Users/vagrant/Data/buildtrees/bitserializer/src/0.10-71fe0ca140 -DBUILD_CPPRESTJSON_ARCHIVE=ON -DBUILD_RAPIDJSON_ARCHIVE=ON -DBUILD_PUGIXML_ARCHIVE=ON -DBUILD_RAPIDYAML_ARCHIVE=ON -DCMAKE_MAKE_PROGRAM=/Users/vagrant/Data/downloads/tools/ninja-1.10.0-osx/ninja -DCMAKE_SYSTEM_NAME=Darwin -DBUILD_SHARED_LIBS=OFF -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=/Volumes/data/work/1/s/scripts/toolchains/osx.cmake -DVCPKG_TARGET_TRIPLET=x64-osx -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=external -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=/Volumes/data/work/1/s/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=/Volumes/data/work/1/s -D_VCPKG_INSTALLED_DIR=/Users/vagrant/Data/installed -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/Users/vagrant/Data/packages/bitserializer_x64-osx/debug
    Working Directory: /Users/vagrant/Data/buildtrees/bitserializer/x64-osx-dbg
    Error code: 1
    See logs for more information:
      /Users/vagrant/Data/buildtrees/bitserializer/config-x64-osx-dbg-out.log
      /Users/vagrant/Data/buildtrees/bitserializer/config-x64-osx-dbg-err.log
```

PavelKisliak commented:
Which options do we have? May be skip build on MacOS for main **bitserializer**?
In future, we can remove this when the CI will be fixed or when we able to remove aliases (issue  #11482).

PhoebeHui commented:
@PavelKisliak, I think we can skip Arm-UWP, arm64-window and OSX  for main bitserializer in baseline for unblocking CI testing.

PhoebeHui commented on scripts/ci.baseline.txt:
```diff
@@ -125,9 +125,9 @@ berkeleydb:arm-uwp=fail
-bitserializer-rapidyaml:x64-osx=fail
+bitserializer-rapidyaml:arm-uwp=skip
+bitserializer-rapidyaml:arm64-windows=skip
+bitserializer-rapidyaml:x64-osx=skip
```
```suggestion
bitserializer-rapidyaml:x64-osx=skip
bitserializer:arm-uwp=skip
bitserializer:arm64-windows=skip
bitserializer:x64-osx=skip
```

PhoebeHui committed:
Update scripts/ci.baseline.txt

BillyONeal commented:
@PhoebeHui I removed depends:vm-update because I don't see any changes to VM deployment scripts here. (I'm tagging you in case you meant to add some other label)
